drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright (c) 2022 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+#include "adc.h"
+#include <zephyr/dt-bindings/adc/b91-adc.h>
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_b91);
+
+#define DT_DRV_COMPAT telink_b91_adc
+
+/* ADC B91 defines */
+#define SIGN_BIT_POSITION          (13)
+#define AREG_ADC_DATA_STATUS       (0xf6)
+#define ADC_DATA_READY             BIT(0)
+
+#define CASE_DT_GPIO_TO_HAL_GPIO(hal_gpio) \
+	case DT_##hal_gpio:			   \
+		adc_pin = hal_gpio;		   \
+		break;
+
+#define CASE_ACK_TIME_TO_HAL_SAMPLE_CYC(n) \
+	case ADC_ACQ_TIME(ADC_ACQ_TIME_TICKS, n):  \
+		sample_cycl = ADC_SAMPLE_CYC_##n;  \
+		break;
+
+/* B91 ADC driver data */
+struct driver_data {
+	struct adc_context ctx;
+	int16_t *buffer;
+	int16_t *repeat_buffer;
+	uint8_t differential;
+	uint8_t resolution_divider;
+	struct k_sem acq_sem;
+	struct k_thread thread;
+
+	K_THREAD_STACK_MEMBER(stack, CONFIG_ADC_B91_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+static struct driver_data m_data = {
+	ADC_CONTEXT_INIT_TIMER(m_data, ctx),
+	ADC_CONTEXT_INIT_LOCK(m_data, ctx),
+	ADC_CONTEXT_INIT_SYNC(m_data, ctx),
+};
```

Comment:
please make this part of device data

Reply:
@gmarull,
Sorry, I did not get the requested changes.
Could you please specify what exactly I have to change here?

Reply:
Done.